asv/benchmarks.py

```diff
@@ -462,26 +468,67 @@ def run_benchmarks(self, env, show_stderr=False, quick=False, profile=False,
 
             - `result`: The numeric value of the benchmark (usually
               the runtime in seconds for a timing benchmark), but may
-              be an arbitrary JSON data structure. For parameterized tests, 
+              be an arbitrary JSON data structure. For parameterized tests,
               this is a dictionary with keys 'params' and 'result', where
               the value of 'params' contains a list of lists of parameter values,
               and 'result' is a list of results, corresponding to itertools.product
-              iteration over parameters. 
+              iteration over parameters.
               Set to `None` if the benchmark failed.
 
             - `profile`: If `profile` is `True`, this key will exist,
               and be a byte string containing the cProfile data.
         """
         log.info("Benchmarking {0}".format(env.name))
-        with log.indent():
-            times = {}
-            benchmarks = sorted(list(six.iteritems(self)))
-            for name, benchmark in benchmarks:
-                if skip and name in skip:
-                    continue
-                times[name] = run_benchmark(
-                    benchmark, self._benchmark_dir, env, show_stderr=show_stderr,
-                    quick=quick, profile=profile)
+
+        cache_file = tempfile.NamedTemporaryFile(suffix='dbm', delete=False)
+        cache_file.close()
+        # We have to actually delete the file here, since the database
+        # type can not be determined from a zero-length file.  We
+        # can't create the database in the master process, since the
+        # file formats available may vary between Python environments.
+        os.remove(cache_file.name)
+        try:
+            with log.indent():
+                benchmarks = sorted(list(six.iteritems(self)))
+
+                setup_caches_done = set()
+                for name, benchmark in benchmarks:
+                    key = benchmark.get('setup_cache_key')
+                    if key is not None and key not in setup_caches_done:
+                        log.info('Caching {0}'.format(key))
+                        setup_caches_done.add(key)
+                        env.run(
+                            [BENCHMARK_RUN_SCRIPT, 'setup_cache',
+                             self._benchmark_dir, name,
+                             cache_file.name],
+                            dots=False, display_error=True,
+                            return_stderr=True)
```

Comment:
Agreed about the failure.  I'll fix that.

I do all of this upfront to be careful about spawning as few processes as possible (by tracking the `setup_caches_done` set).  That could probably done inside of `run_benchmark` but would be more cumbersome.  Is there any advantage to that?

Agree that it needs to track the skip list.  I might just pre-filter the benchmark list.